discovery_engine/lib/src/ffi/types/document/feedback.dart

```diff
@@ -0,0 +1,65 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import 'dart:ffi' show Pointer;
+
+import 'package:xayn_discovery_engine/src/domain/models/document.dart'
+    show DocumentFeedback;
+import 'package:xayn_discovery_engine/src/ffi/genesis.ffigen.dart'
+    show RustUserReaction;
+import 'package:xayn_discovery_engine/src/ffi/load_lib.dart' show ffi;
+
+extension DocumentFeedbackFfi on DocumentFeedback {
+  void writeNative(final Pointer<RustUserReaction> place) {
+    var asInt = 128;
+    switch (this) {
+      case DocumentFeedback.neutral:
+        asInt = 0;
+        break;
+      case DocumentFeedback.positive:
+        asInt = 1;
+        break;
+      case DocumentFeedback.negative:
+        asInt = 2;
+        break;
+    }
+    final ok = ffi.init_user_reaction_at(place, asInt);
+    if (ok == 0) {
+      throw ArgumentError.value(
+        this,
+        'DocumentFeedback',
+        'unsupported DocumentFeedback variant',
+      );
+    }
+  }
+
+  static DocumentFeedback readNative(
+    final Pointer<RustUserReaction> place,
+  ) {
+    final reaction = ffi.get_user_reaction(place);
+    switch (reaction) {
+      case 0:
+        return DocumentFeedback.neutral;
+      case 1:
+        return DocumentFeedback.positive;
+      case 2:
+        return DocumentFeedback.negative;
+    }
+    throw ArgumentError.value(
```

Comment:
since we have to do the mapping on the dart side anyways, why don't we simply put a `#[repr(u8)]` (or a `#[repr(IntRepr)]` if we want to stick with the alias) at the rust `UserReaction` (which only has fieldless variants anyways) and export it in the ffi? this would give us a bunch of named constants which we could use in the mappings, and the ffi functions could take a `UserReaction` value directly instead of a `*mut UserReaction` pointer (which would make the alloc/init/drop ffi unnecessary).

Reply:
You still need a `*mut UserReaction` as outside of testing you are basically always write the enum to a field.

Though you could merge the field addr + getter/setter, but that is less reusable/composable and less consistent with he other fields.

The `alloc_`/`drop_` tests are used by dart tests only, so we are not really saving anything there.

Reply:
I didn't make `UserReaction` non-opaque, as then cbindgen has to spider additional crates.

Reply:
I originally had it as `repr(u8)` but changed it away from it. I don't remember why. But one reason could have been that I didn't want to add ffi-only int representations to the enum :thinking: . Not sure.

Reply:
i see, thx for the explanation. i'm still not super happy with the solution, but at least for the sake of consistency it's ok for me.